plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/s3/S3FileSystemExchangeStorage.java

```diff
@@ -0,0 +1,565 @@
+    {
+        S3ClientBuilder clientBuilder = S3Client.builder()
+                .credentialsProvider(credentialsProvider)
+                .overrideConfiguration(overrideConfig);
+
+        if (region != null) {
+            clientBuilder = clientBuilder.region(region);
+        }
+
+        return clientBuilder.build();
+    }
+
+    private S3AsyncClient createS3AsyncClient(AwsCredentialsProvider credentialsProvider, ClientOverrideConfiguration overrideConfig)
+    {
+        S3AsyncClientBuilder clientBuilder = S3AsyncClient.builder()
+                .credentialsProvider(credentialsProvider)
+                .overrideConfiguration(overrideConfig);
+
+        if (region != null) {
+            clientBuilder = clientBuilder.region(region);
+        }
+
+        return clientBuilder.build();
+    }
+
+    private static class S3ExchangeStorageWriter
+            implements ExchangeStorageWriter
+    {
+        private final S3Client s3Client;
+        private final S3AsyncClient s3AsyncClient;
+        private final String bucketName;
+        private final String key;
+        private final int partSize;
+        private final Optional<SecretKey> secretKey;
+
+        private int currentPartNumber;
+        private Optional<String> uploadId = Optional.empty();
+        private final List<CompletableFuture<CompletedPart>> uploadFutures = new ArrayList<>();
+
+        public S3ExchangeStorageWriter(S3Client s3Client, S3AsyncClient s3AsyncClient, String bucketName, String key, int partSize, Optional<SecretKey> secretKey)
+        {
+            this.s3Client = requireNonNull(s3Client, "s3Client is null");
+            this.s3AsyncClient = requireNonNull(s3AsyncClient, "s3AsyncClient is null");
+            this.bucketName = requireNonNull(bucketName, "bucketName is null");
+            this.key = requireNonNull(key, "key is null");
+            this.partSize = partSize;
+            this.secretKey = requireNonNull(secretKey, "secretKey is null");
+        }
+
+        @Override
+        public ListenableFuture<Void> write(Slice slice)
+                throws IOException
+        {
+            // skip multipart upload if there would only be one part
+            if (slice.length() < partSize && uploadId.isEmpty()) {
+                PutObjectRequest.Builder putObjectRequestBuilder = PutObjectRequest.builder()
+                        .bucket(bucketName)
+                        .key(key);
+                S3RequestUtil.configureEncryption(secretKey, putObjectRequestBuilder);
+
+                try {
+                    s3Client.putObject(putObjectRequestBuilder.build(),
+                            // avoid extra memory copy
+                            fromContentProvider(() -> new ByteArrayInputStream(slice.getBytes()), slice.length(), Mimetype.MIMETYPE_OCTET_STREAM));
+                    return immediateVoidFuture();
```

Comment:
Can it be done in an asynchronous way to avoid blocking?

Reply:
I think doing this async will break semantic correctness. We have to finish writing the data files before creating the COMMIT file, right?